### ISSUE TITLE
Inconsistent conversions to/from zoo with zero-length coredata

### ISSUE THREAD

**Issue by joshuaulrich** [joshuaulrich/xts#228]
Converting a zero-length object to zoo and back to xts loses the dim and dimname attributes.

```r
# zero-length zoo
z <- zoo(matrix(numeric(0), 3, 3)[0,], numeric(0))

# can't convert to xts
as.xts(z)
### Error in xts(coredata(x), order.by = order.by, frequency = frequency,  :
###   order.by requires an appropriate time-based object

# zero-length xts
x <- .xts(matrix(numeric(0), 3, 3)[0,], numeric(0))

# can convert to zoo...
zx <- as.zoo(x)

# ...but can't convert back to xts
xzx <- as.xts(zx)
### Error in xts(coredata(x), order.by = order.by, frequency = frequency,  :
###   order.by requires an appropriate time-based object
```

This is partially caused by `as.zoo.xts()` setting coredata to `NULL`. It might also be related to the fix for #223.

**Comment by joshuaulrich** [comment]
It's not an issue that `as.xts(z)` doesn't work because `z` doesn't have a time-based index. It works if the index is time-based.

```r
z <- zoo(matrix(numeric(0), 3, 3)[0,], .POSIXct(numeric(0)))
xz <- as.xts(z)
```

We have to change `zoo:::as.zoo.xts()` in order for `all.equal(x, xzx)` to be TRUE. It needs to only set `y <- NULL` if it's zero-length and has no dims.

The patch is:

```diff
diff --git a/pkg/zoo/R/as.zoo.R b/pkg/zoo/R/as.zoo.R
index d1a8c99..0f5dc42 100644
--- a/pkg/zoo/R/as.zoo.R
+++ b/pkg/zoo/R/as.zoo.R
@@ -52,7 +52,7 @@ as.zoo.timeSeries <- function(x, ...) {
 
 as.zoo.xts <- function(x, ...) {
   y <- coredata(x)
-  if (length(y) == 0) {
+  if (length(y) == 0 && is.null(dim(y))) {
     y <- NULL
   }
   zoo(y, order.by = index(x), ...)
```

**Comment by joshuaulrich** [comment]
This is fixed in zoo >= 1.8-12.

**Comment by zeileis** [comment]
Thanks, much appreciated.
